src/infinite-scroll/index.ts

```diff
@@ -0,0 +1,356 @@
+          this.state.scroller.root.addEventListener('scroll', this.scroll);
+        }
+      } else if (this.state.oneTime) {
+        this.state.scroller.root.addEventListener('scroll', this.scroll);
+      }
+
+      this.state = { ...this.state, ...state };
+    }
+
+    if (this.state.setScroll) this.setScroll();
+  }
+
+  setScroll = () => {
+    const imgs = <any>this.state.wrapper.querySelectorAll('img') || [];
+    const pageSize = this.state.pageSize(this.flux.store.getState());
+    let count = 0;
+
+    const imgsLoaded = (total, size, totalImages) =>
+      count === size || count === imgs.length;
+
+    if (imgs.length > 0) {
+      for (let i = 0; i < pageSize; i++) {
+        imgs[i].onload = () => {
+          count++;
+          if (imgsLoaded(count, pageSize, imgs.length)) {
+            this.maintainScrollTop(this.state.rememberScroll);
+            this.state.scroller.root.addEventListener('scroll', this.scroll);
+          }
+        };
+      }
+      // need to still add scrollTop & scroll listener if imgs don't load
+      setTimeout(() => {
+        if (!imgsLoaded(count, pageSize, imgs.length)) {
+          this.maintainScrollTop(this.state.rememberScroll);
+          this.state.scroller.root.addEventListener('scroll', this.scroll);
+        }
+      }, 500);
+    }
+
+    this.state = { ...this.state, setScroll: false };
+  }
+
+  updateProducts = () => {
+    const items = this.setProducts();
+
+    this.state = {
+      ...this.state,
+      elItems: this.state.wrapper.children,
+      firstEl: items[0],
+      lastEl: items[items.length - 1],
+      getPage: false,
+      oneTime: true,
+    };
+  }
+
+  setProducts = (products?: Store.ProductWithMetadata[]) => {
+    // TODO: put the products at the end if they're forward, at the beginning if they're backward
+    let items;
+    if (products) {
+      if (products[0].index > this.state.items[this.state.items.length - 1].index) {
+        items = [...this.state.items, ...products.map(this.productTransformer)];
+        this.set(<any>{
+          items,
+          prevExists: items[0].index !== 1,
+          moreExists: this.state.recordCount(this.flux.store.getState()) !== items[items.length - 1].index,
+        });
+      } else if (products[products.length - 1].index < this.state.items[0].index) {
+        const pageSize = this.state.pageSize(this.flux.store.getState());
+        const rememberScroll = this.calculateOffset(pageSize) + this.state.scroller.root.scrollTop;
+        items = [...products.map(this.productTransformer), ...this.state.items];
+        this.set(<any>{
+          ...this.state,
+          items,
+          setScroll: true,
+          rememberScroll,
+          prevExists: items[0].index !== 1,
+          moreExists: this.state.recordCount(this.flux.store.getState()) !== items[items.length - 1].index,
+        });
+        this.state.scroller.root.removeEventListener('scroll', this.scroll);
+      }
+    } else {
```

Comment:
It seems that if statement and else statement don't share common code whether you pass in products or not, why not break it into two different functions?